enhancements/installer/inputs-from-stdin.md

```diff
@@ -0,0 +1,182 @@
+  - "@copejon"
+reviewers:
+  - "@fzdarsky"
+approvers:
+  - "@crawford"
+  - "@staebler"
+  - "@dgoodwin"
+creation-date: 2020-12-24
+last-updated: 2021-01-13
+status: provisional
+---
+
+# Inputs from STDIN
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+In many occasions, users need to customize the install-config target and/or the manifests generated by the OpenShift Installer in a reproducible manner. The most popular tool to allow this type of customization is “Kustomize”, and it is already well-known by the Kubernetes ecosystem. The result of any “Kustomization” is sent to standard output as a stream of manifests.
+
+This OpenShift enhancement proposes to allow the installer to read a stream from STDIN on the main targets that consume inputs.
+
+
+## Motivation
+
+Infrastructure-as-Code (e.g. GitOps) is an operational best practice for provisioning and managing large numbers of clusters and services. A typical IaC workflow involves pulling Kubernetes resource manifest templates of a specific version from a `git` repo, rendering those templates for a given cluster instance, and then running `openshift-install` (resp. `oc apply`) on the rendered manifests to install (resp. update) a cluster.
+
+Pulling and rendering manifest templates is conveniently done via `kustomize` - which is particularly suited for an IaC workflow - but other tools like `helm` and `jsonnet` have their followers, too. These tools have in common that they render templates into a stream of manifests sent to `stdout`.
+
+As `openshift-install` currently only supports reading manifest files from the file system and requires files to follow a specific naming and directory structure convention, IaC workflows become cumbersome and often involve users writing their own Bash scripts or Ansible playbooks to automate the workflows. By giving the installer the ability to read from `stdin`, we will enable a simplified workflow pattern via pipelines while still being capable of scaling with the complexity of users’ needs.
+
+### Goals
+
+Enable the Openshift Installer to support pipelined workflows by reading inputs from STDIN as an alternative to reading from the filesystem.
+
+### Non-Goals
+
+- Change the existing workflow to deploy an OpenShift cluster.
+- Couple the Openshift Installer with Kustomize or other manifest tooling
+
+
+## Proposal
+
+Modify the OpenShift Installer to accept inputs from STDIN for all targets. For example: 
+
+- Create manifests from customized install-config.yaml: `kustomize build github.com/myorg/myBlueprint/install-config-kustomization | openshift-install create manifests -f -`
```

Comment:
The expectation is that the installer can read manifests from stdin like it currrently can read from a `--dir`. The current proposal is to express this to use `-f -` as argument, but if something like `--stdin` makes more sense, that'd be perfectly fine.

Just like the installer doesn't know a priori whether the `--dir` contains an install-config.yaml or additional manifests, it wouldn't know reading from `--stdin`. The assumption we've made is that once the resources read from stdin are mapped back into the same data structure when reading from `--dir`, the following logic would be identical.

Re `openshift-install --kustomization create manifests --dir=$TMPDIR`: This could work and was in fact part of an earlier proposal. However, we've been given the guidance by the architects _not_ to propose an approach that would make the installer opinionated on a given templating tool or even to expect kustomize libs to be added to the installer. This is why we're proposing the stdin approach now.

Reply:
There is more in the assets directory than just manifests. Is your expectation that **everything** in the assets directory is going to be sent in via STDIN? Or is, for example, the asset store going to remain in the assets directory?

You are correct that the installer does not know a priori whether the assets director contains and install-config.yaml or additional manifests. The installer builds the asset graph up to the targets desired in the command and tries to load from the asset directory whatever assets are needed. The installer relies on the names of files to ascertain whether the files for a particular asset are on disk. It is not clear to me how the "mapp[ing] back into the same data structure" is supposed to work without the file names. Let's take the ignition config assets as an example. How is the installer expected to differentiate between the master ignition config and the worker ignition config when read from STDIN?

Re `openshift-install --kustomization create manifests --dir=$TMPDIR`: You can ignore the use of the `--kustomization` flag here. That was lifted from on of the use cases later in the document. It has no bearing on the behavior of the installer as it relates to input.